### PR TITLE
api/v1: Expose rules and alerts

### PR DESCRIPTION
**Pull request by mxinden**
This PR builds on top of @mg03 work and is a follow up to https://github.com/prometheus/prometheus/pull/4012:
- Addressing outstanding comments on https://github.com/prometheus/prometheus/pull/4012
- Adjusting to Prometheus coding style
- Differentiating between alerting and recording rules
- ...

The first commit contains @mg03 original changes rebased to current `master`. The second commit contains all follow up changes by me.

Let me know what you think.

//CC @brancz @kyoto

### CODE REVIEW

File: rules/alerting.go [hunk 1]
```diff
@@ -131,6 +131,26 @@ func (r *AlertingRule) Name() string {
...
+func (r *AlertingRule) Query() promql.Expr {
+	return r.vector
+}
+
+// Duration returns the hold duration of the alert.
+func (r *AlertingRule) Duration() time.Duration {
+	return r.holdDuration
+}
+
+// Labels returns the labels of the alert.
```
Comment (brian-brazil) [inline rules/alerting.go]: alerting rule

alert labels are a bit different due to templating

File: rules/recording.go [hunk 2]
```diff
@@ -52,6 +52,16 @@ func (rule *RecordingRule) Name() string {
 	return rule.name
 }
 
+// Query returns the rule query expression
```
Comment (brian-brazil) [inline rules/recording.go]: Full stop

File: web/api/v1/api.go [hunk 3]
```diff
@@ -574,6 +585,130 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
 	return ams, nil, nil
 }
 
+// AlertDiscovery has info for all alerts
+type AlertDiscovery struct {
+	AlertGroups []*AlertGroup `json:"alertGroups"`
+}
+
+// Alert has info for a alert
```
Comment (brian-brazil) [inline web/api/v1/api.go]: an alert

File: web/api/v1/api.go [hunk 4]
```diff
@@ -574,6 +585,130 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
...
+type Alert struct {
+	Labels      labels.Labels `json:"labels"`
+	Annotations labels.Labels `json:"annotations"`
+	State       string        `json:"state"`
+	ActiveAt    *time.Time    `json:"activeAt,omitempty"`
```
Comment (brian-brazil) [inline web/api/v1/api.go]: We might want to be a little careful with this one, its semantics may change over time.

File: web/api/v1/api.go [hunk 5]
```diff
@@ -574,6 +585,130 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
...
+}
+
+// AlertGroup has info for alerts part of a group
+type AlertGroup struct {
+	Name        string        `json:"name"`
+	Query       string        `json:"query"`
+	Duration    string        `json:"duration"`
+	Annotations labels.Labels `json:"annotations,omitempty"`
```
Comment (brian-brazil) [inline web/api/v1/api.go]: No labels?

File: web/api/v1/api.go [hunk 6]
```diff
@@ -574,6 +585,130 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
...
+// Alert has info for a alert
+type Alert struct {
+	Labels      labels.Labels `json:"labels"`
+	Annotations labels.Labels `json:"annotations"`
+	State       string        `json:"state"`
+	ActiveAt    *time.Time    `json:"activeAt,omitempty"`
+}
+
+// AlertGroup has info for alerts part of a group
+type AlertGroup struct {
```
Comment (brian-brazil) [inline web/api/v1/api.go]: An alert group is something we have in the alertmanager. This is the output of an alerting rule.

File: web/api/v1/api.go [hunk 7]
```diff
@@ -574,6 +585,130 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
...
+}
+
+func (api *API) alerts(r *http.Request) (interface{}, *apiError, func()) {
+	alertingRules := api.rulesRetriever.AlertingRules()
+	var alertGroups []*AlertGroup
+
+	for _, alertingRule := range alertingRules {
+		alertGroup := &AlertGroup{
+			Name:        alertingRule.Name(),
+			Query:       fmt.Sprintf("%v", alertingRule.Query()),
```
Comment (brian-brazil) [inline web/api/v1/api.go]: Do you really need the sprintf here?

**Comment by mxinden** [comment]
> The docs need updating.

@brian-brazil Where would you like the two endpoints to be documented? `querying/api.md` seems to only be related to the `/query` api endpoint.

In addition, should I add a comment to `stability.md` mentioning that this endpoint is subject to change even though it is part of the v1 API?

**Comment by brian-brazil** [comment]
api.md is where we put everything. A comment in api.md is sufficient.

File: web/api/v1/api.go (hunk 6 above)
Comment (mxinden) [reply web/api/v1/api.go]: Good point. Should I rename this to `AlertingRule` or should we flatten (/denormalize) the endpoint and expose a list of alerts right away?

Comment (brian-brazil) [reply web/api/v1/api.go]: Flattening is probably easiest.

File: web/api/v1/api.go (hunk 5 above)
Comment (mxinden) [reply web/api/v1/api.go]: Good point. Added below. :heavy_check_mark:

**Comment by brian-brazil** [comment]
That's insufficient to identify a rule, you can have multiple of the same name in a group.

> Tying an alert and an alerting rule together seems like a valid use case to me.

Noone has ever asked for it that I can recall.

> Adding filename and rule group name to each alert exposed on the /alerts endpoint.

That could get rather large size wise.

> Internally adding a filename and groupname label to each alert, as we already do with the alertname.

This would be a breaking change for both ALERTs and alerting generally. From a semantic standpoint it doesn't make sense, you don't care where the alert came from just that it is firing.

**Comment by brancz** [comment]
The ability to tell which rule produced which alert seems reasonable to me (and is a use case we have).

How about we don't attach that information by default but make it available via alert relabeling, similar to optional discovery information as `__meta` labels? If I understand you correctly @brian-brazil your comment on this being a breaking change only refers to the set of labels that make up an alert (which I agree we absolutely cannot break the default behavior).

**Comment by mxinden** [comment]
What are your thoughts on option number three?

> Nesting all active alerts inside the `/rules` endpoint in each respective rules object.

This would result in a rather deep nesting, but enables us to tie each alert to its alerting rule.

**Comment by brian-brazil** [comment]
> How about we don't attach that information by default but make it available via alert relabeling, similar to optional discovery information as __meta labels?

I don't see how that helps, this should never hit the alertmanager.

**Comment by brancz** [comment]
> I don't see how that helps, this should never hit the alertmanager.

Hmm I see your point. It's easy enough to identify the source Prometheus via external labels anyways. And then it would only need to be possible to have Prometheus tell which rule produced an alert (but never sending this information to Alertmanager). I think ideally it would be nice if (within Prometheus) we could tell the alert(s) a specific rule is creating and vice versa which alert was produced by which rule.

Returning the respective rule to an alert could be optional, if we're worried about the amount of data returned.

**Comment by mxinden** [comment]
Flattened `/alerts` endpoint and nested active alerts of each rule into the `/rules` endpoint.

You find a sample endpoint payload in the `api.md` changes.

File: web/api/v1/api.go [hunk 8]
```diff
@@ -574,6 +585,120 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
 	return ams, nil, nil
 }
 
+// AlertDiscovery has info for all active alerts.
+type AlertDiscovery struct {
+	ActiveAlerts []*Alert `json:"activeAlerts"`
```
Comment (mxinden) [inline web/api/v1/api.go]: Happy to rename this back to `Alerts`. `ActiveAlerts` gives the user a hint, that resolved alerts are not included.

File: web/api/v1/api.go [hunk 9]
```diff
@@ -574,6 +585,120 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
...
+	Interval       time.Duration   `json:"interval"`
+}
+
+type alertingRule struct {
+	Name         string        `json:"name"`
+	Query        string        `json:"query"`
+	Duration     time.Duration `json:"duration"`
+	Labels       labels.Labels `json:"labels"`
+	Annotations  labels.Labels `json:"annotations"`
+	ActiveAlerts []*Alert      `json:"activeAlerts"`
```
Comment (mxinden) [inline web/api/v1/api.go]: Same as above: Happy to rename this to `Alerts`. `ActiveAlerts` gives the user a hint, that resolved alerts are not included.

File: web/api/v1/api_test.go [hunk 10]
```diff
@@ -98,6 +101,74 @@ func (t testAlertmanagerRetriever) DroppedAlertmanagers() []*url.URL {
...
+	)
+	rule2 := rules.NewAlertingRule(
+		"test_metric4",
+		expr2,
+		time.Second,
+		labels.Labels{},
+		labels.Labels{},
+		log.NewNopLogger(),
+	)
+	// TODO: Should also contain test for recording rule?
```
Comment (simonpasquier) [inline web/api/v1/api_test.go]: Is it still valid?

**Comment by mxinden** [comment]
@simonpasquier @brian-brazil @brancz thanks for your comments. Would you mind taking another look?

File: docs/querying/api.md [hunk 11]
```diff
@@ -363,6 +363,91 @@ $ curl http://localhost:9090/api/v1/targets
...
+```
+
+```json
+$ curl http://localhost:9090/api/v1/rules
+
+{
+    "data": {
+        "groups": [
+            {
+                "alertingRules": [
```
Comment (brian-brazil) [inline docs/querying/api.md]: There is an ordering across all rules, the alert recording and recording rules need to be in the same list.

File: docs/querying/api.md [hunk 12]
```diff
@@ -363,6 +363,91 @@ $ curl http://localhost:9090/api/v1/targets
...
+                        ],
+                        "annotations": {},
+                        "duration": 0,
+                        "labels": {},
+                        "name": "my-alert",
+                        "query": "vector(1)"
+                    }
+                ],
+                "file": "",
+                "interval": 60000000000,
```
Comment (brian-brazil) [inline docs/querying/api.md]: we use seconds

File: web/api/v1/api.go (hunk 4 above)
Comment (brian-brazil) [reply web/api/v1/api.go]: Yes. Users also depend on these values in ways that aren't correct, these are implementation details more than anything.

File: docs/querying/api.md (hunk 11 above)
Comment (mxinden) [reply docs/querying/api.md]: But giving up the `alertingRule` and recordingRule` bucket prevents us from differentiating the two, thereby we are not able to expose details like `Duration`, `Annotations`, active `Alerts`, ...

In which use case is the ordering of rules important?

Comment (brian-brazil) [reply docs/querying/api.md]: In any use case where there are interdependent rules and alerts.

File: web/api/v1/api.go [hunk 13]
```diff
@@ -574,6 +585,121 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
...
+		)
+	}
+
+	res := &AlertDiscovery{Alerts: alerts}
+
+	return res, nil, nil
+}
+
+func rulesAlertsToAPIAlerts(rulesAlerts []*rules.Alert) []*Alert {
+	apiAlerts := []*Alert{}
```
Comment (simonpasquier) [inline web/api/v1/api.go]: super minor but it could be `apiAlerts := make([]*Alert{}, len(rulesAlerts))`

File: docs/querying/api.md (hunk 11 above)
Comment (simonpasquier) [reply docs/querying/api.md]: I think @brian-brazil has a point. I see several options.

1) add a `type` field that would be `alerting` or `recording`.

```json
{
    "data": {
        "groups": [
            {
                "rules": [
                    {
                        "alerts": [
                            {
                                "activeAt": "2018-07-04T20:27:12.60602144+02:00",
                                "annotations": {},
                                "labels": {
                                    "alertname": "my-alert"
                                },
                                "state": "firing",
                                "value": 1
                            }
                        ],
                        "annotations": {},
                        "duration": 0,
                        "labels": {},
                        "name": "my-alert",
                        "query": "vector(1)",
                        "type": "alerting"
                    }, {
                        "duration": 0,
                        "labels": {},
                        "name": "my-record",
                        "query": "vector(0)",
                        "type": "recording"
                    }
                ],
                "file": "",
                "interval": 60,
                "name": ""
            }
        ]
    },
    "status": "success"
}
```
2) Keep your proposal and add an `index` field.

```json
{
    "data": {
        "groups": [
            {
                "alertingRules": [
                    {
                        "alerts": [
                            {
                                "activeAt": "2018-07-04T20:27:12.60602144+02:00",
                                "annotations": {},
                                "labels": {
                                    "alertname": "my-alert"
                                },
                                "state": "firing",
                                "value": 1
                            }
                        ],
                        "annotations": {},
                        "duration": 0,
                        "index": 0,
                        "labels": {},
                        "name": "my-alert",
                        "query": "vector(1)"
                    }
                ],
                "file": "",
                "interval": 60,
                "name": "",
                "recordingRules": [{
                        "duration": 0,
                        "index": 1,
                        "labels": {},
                        "name": "my-record",
                        "query": "vector(0)",
                        "type": "recording"
                    }
                ]
            }
        ]
    },
    "status": "success"
}
```

I'd go with the first option as alerting rules are a superset of recording rules.

**Comment by mxinden** [comment]
@brian-brazil @simonpasquier Thanks for the hint on the ordering. Both alerting and recording rules are exposed in the same `rules` array now, differentiating the two via a `type` property.

Further comments / suggestions?

File: docs/querying/api.md [hunk 14]
```diff
@@ -363,6 +363,98 @@ $ curl http://localhost:9090/api/v1/targets
...
+                        "annotations": {},
+                        "duration": 0,
+                        "labels": {},
+                        "name": "my-alerting-rule",
+                        "query": "vector(1)",
+                        "type": "alerting"
+
+                    },
+                    {
+                        "labels": {},
```
Comment (brian-brazil) [inline docs/querying/api.md]: Should we have an omitempty on this? Labels aren't typical on recording rules, and having this here could be a bit confusing.

File: docs/querying/api.md [hunk 15]
```diff
@@ -363,6 +363,98 @@ $ curl http://localhost:9090/api/v1/targets
...
+                        "name": "my-recording-rule",
```
Comment (brian-brazil) [inline docs/querying/api.md]: Can we use a valid metric name?

File: docs/querying/api.md [hunk 16]
```diff
@@ -363,6 +363,98 @@ $ curl http://localhost:9090/api/v1/targets
...
+        "groups": [
+            {
+                "rules": [
+                    {
+                        "alerts": [
+                            {
+                                "activeAt": "2018-07-04T20:27:12.60602144+02:00",
+                                "annotations": {},
+                                "labels": {
+                                    "alertname": "my-alert"
```
Comment (brian-brazil) [inline docs/querying/api.md]: This is inconsistent with the `name`. MyAlertingRule is more typical.

File: web/api/v1/api.go [hunk 17]
```diff
@@ -574,6 +585,132 @@ func (api *API) alertmanagers(r *http.Request) (interface{}, *apiError, func())
...
+type rule interface{}
+
+type alertingRule struct {
+	Name        string        `json:"name"`
+	Query       string        `json:"query"`
+	Duration    time.Duration `json:"duration"`
+	Labels      labels.Labels `json:"labels"`
+	Annotations labels.Labels `json:"annotations"`
+	Alerts      []*Alert      `json:"alerts"`
+	// Type of an alertingRule is always "alerting"
```
Comment (brian-brazil) [inline web/api/v1/api.go]: Full stop

**Comment by mxinden** [comment]
@brian-brazil thanks for the comments! I very much appreciated the help.

I included the following changes:https://prometheus.io/docs/prometheus/latest/configuration/
- Updated the examples now derived from the [Prometheus docs](https://prometheus.io/docs/prometheus/latest/configuration/).
- Add `omitempty` struct tag to `Labels` of a recording rule.
- Change `alertingRule.Duration` from `time.Duration` to `float64` using base unit seconds.

Let me know if you have further suggestions.

In addition, we need to think about how we want to proceed in regards to the DCO check. As the first commit is not authored by me, I don't want to sign it. I would suggest a PR approval by @mg03. What do you think @brian-brazil?

**Comment by brian-brazil** [comment]
That sounds reasonable to me. You will need to adjust the commit message.

**Comment by mxinden** [comment]
> You will need to adjust the commit message.

@brian-brazil Do you suggest me to add `Signed-off-by: Max Leonard Inden <IndenML@gmail.com>` to the first commit authored by @mg03? That sounds a bit odd to me.

**Comment by brian-brazil** [comment]
No, signed off by mg03's email if they approve.

**Comment by mg03** [comment]
@mxinden @brian-brazil : I approve. Thanx.

Let me know if anything else is required

**Comment by mxinden** [comment]
Thanks @mg03!

I have adjusted your commit message:

```git
commit 31f8ca0dfb497f12404aba3108440f79b3015659
Author: mg03 <mgeng03@gmail.com>
Date:   Sun Mar 25 09:50:34 2018 -0700

    api v1 alerts/rules json endpoint
    
    Signed-off-by: mg03 <mgeng03@gmail.com>
```

@brian-brazil @simonpasquier anything else missing here?

**Comment by simonpasquier** [comment]
LGTM thanks @mg03 and @mxinden for the hard work!

**Comment by mxinden** [comment]
@brian-brazil looks good to you as well? Can we include this in `v2.4.0`?

**Comment by brian-brazil** [comment]
:+1: 

Everything in master gets included in the next minor release.

**Comment by mxinden** [comment]
Thanks @simonpasquier @brian-brazil and @mg03.

**Comment by danny311296** [comment]
Is there any way to get a lost of all times of alert occurrences from the API.
For example it an alert1 occurred at time t1,t2 and t3,
some sort of API that returns the json with the various timestamps at which rules where triggered.
My use case requires to use prometheus as a time series memory store with a flask backend to handle connecting to prometheus, querying, plotting alerts, ect

**Pull request merged**